scipy/stats/_rcont/_rcont.c

```diff
@@ -0,0 +1,242 @@
+/**
+  This file contains two algorithms written in C to generate random two-way
+  tables. The algorithms rcont1 and rcont2 originate from Boyett and Patefield,
+  respectively. For more information, see the docs for each function.
+
+  If you wonder about the spelling, rcont is short for random contingency
+  table. Random contingency table is a bit of a misnomer. The tables generated
+  by these algorithms have no contingency/association between the two
+  variables.
+
+  Author: Hans Dembinski
+*/
+
+#include "_rcont.h"
+#include <math.h>
+
+// helper function to access a 1D array like a C-style 2D array
+double *ptr(double *m, int nr, int nc, int ir, int ic)
+{
+  return m + nc * ir + ic;
+}
+
+// sanity checks for inputs to rcont1 and rcont2,
+// also sets total number of entries
+int rcont_check(double *n, const double *m, int nr, const double *r, int nc, const double *c)
+{
+  if (m == 0 || r == 0 || c == 0 || n == 0)
+    return 1;
+
+  if (nr < 2 || nc < 2)
+    return 2;
+
+  // check sum(r) == sum(c); r[i] >= 0, c[i] >= 0; sum(r) > 0
+  *n = 0;
+  for (int i = 0; i < nc; ++i)
+  {
+    if (!(c[i] >= 0))
+      return 3;
+    *n += c[i];
+  }
+  double n2 = 0;
+  for (int i = 0; i < nr; ++i)
+  {
+    if (!(r[i] >= 0))
+      return 3;
+    n2 += r[i];
+  }
+  if (*n != n2)
+    return 4;
+  if (!(*n > 0))
+    return 5;
+
+  return 0;
+}
+
+/*
+  Call this once to initialize workspace for rcont1.
+
+  The work space must have size N, where N is the total number of entries.
+*/
+void rcont1_init(int *work, int nc, const double *c)
+{
+  for (int i = 0; i < nc; ++i)
+  {
+    int ci = (int)c[i];
+    while (ci--)
+      *work++ = i;
+  }
+}
+
+/*
+  Generate random two-way table with given marginal totals.
+
+  Boyett's shuffling algorithm adapted from AS 144 Appl. Statist. (1979)
+  329-332. The algorithm has O(N) complexity in space and time for
+  a table with N entries in total. The algorithm performs poorly for large N,
+  but is insensitive to the number K of table cells.
+
+  This function uses a work space of size N which must be preallocated and
+  initialized with rcont1_init.
+*/
+void rcont1(double *matrix, int nr, const double *r, int nc, const double *c,
```

Comment:
I can confirm that this exactly follows the FORTRAN code in Boyett's paper.

Reply:
Good to know, although I had not read Boyett's code when I wrote this. I only discovered in hindsight that this is Boyett's algorithm.